cpp/src/arrow/dataset/dataset.h

```diff
@@ -37,11 +38,107 @@ namespace dataset {
 
 using RecordBatchGenerator = std::function<Future<std::shared_ptr<RecordBatch>>()>;
 
+/// \brief Description of a column to scan
+struct FragmentSelectionColumn {
+  /// \brief The path to the column to load
+  FieldPath column_reference;
+  /// \brief The requested type of the column
+  ///
+  /// A format may choose to ignore this field completely.  For example, when
+  /// reading from IPC the reader can just return the column in the data type
+  /// that is stored on disk.  In this case the scan node will cast the column
+  /// to the appropriate type later.
+  ///
+  /// However, some formats may be capable of casting on the fly.  For example,
+  /// when reading from CSV, if we know the target type of the column, we can
+  /// convert from string to the target type as we read.
+  DataType* requested_type;
+};
+/// \brief Instructions for scanning a particular fragment
+///
+/// The fragment scan request is dervied from ScanV2Options.  The main
+/// difference is that the scan options are based on the dataset schema
+/// while the fragment request is based on the fragment schema.
+struct FragmentScanRequest {
+  /// \brief A row filter
+  ///
+  /// The filter expression should be written against the fragment schema.
+  ///
+  /// \see ScanV2Options for details on how this filter should be applied
+  compute::Expression filter = compute::literal(true);
+
+  /// \brief The columns to scan
+  ///
+  /// These indices refer to the fragment schema
+  ///
+  /// Note: This is NOT a simple list of top-level column indices.
+  /// For more details \see ScanV2Options
+  ///
+  /// If possible a fragment should only read from disk the data needed
+  /// to satisfy these columns.  If a format cannot partially read a nested
+  /// column (e.g. JSON) then it must apply the column selection (in memory)
+  /// before returning the scanned batch.
```

Comment:
Hmm, wouldn't we want to keep this responsibility in a more common place?

Reply:
I suppose the scanner can do this "field pruning" just before the batch is sent for evolution.  I'm not entirely sure how the scanner would know which fields need to be pruned though.  The fragment scanner could maybe have a boolean flag whether it will do the pruning or not do the pruning.

Reply:
Could it be treated as part of the 'schema evolution' functionality, since that's already being used to fill in missing fields?

Reply:
It could be.  I think I'd like to leave it for now and see how it looks when we add CSV support for the new scanner.